### PR TITLE
recursive call in user directories for unit files

### PR DESCRIPTION
**Pull request by harikannan512**
#### Does this PR introduce a user-facing change?

Solves #18032 

```release-note
Recursively scans directories for unit files
----------------------
debian@debian:~/Projects/podman$ bin/quadlet -dryrun
quadlet-generator[2453]: No files to parse from [/etc/containers/systemd /usr/share/containers/systemd]

----------------------
debian@debian:~/Projects/podman$ bin/quadlet -dryrun -user
quadlet-generator[2475]: Loading source unit file /home/debian/.config/containers/systemd/test_subdir/sample.container
quadlet-generator[2475]: Loading source unit file /home/debian/.config/containers/systemd/test_subdir/sub_one/sample_two.container
quadlet-generator[2475]: Loading source unit file /home/debian/.config/containers/systemd/test_subdir/sub_two/sample_one.container
---sample.service---
[Unit]
Description=The sleep container
After=local-fs.target
SourcePath=/home/debian/.config/containers/systemd/test_subdir/sample.container
RequiresMountsFor=%t/containers

[X-Container]
Image=registry.access.redhat.com/ubi9-minimal:latest
Exec=sleep 1000

[Install]
# Start by default on boot
WantedBy=multi-user.target default.target

[Service]
Environment=PODMAN_SYSTEMD_UNIT=%n
KillMode=mixed
ExecStop=/usr/local/bin/podman rm -f -i --cidfile=%t/%N.cid
ExecStopPost=-/usr/local/bin/podman rm -f -i --cidfile=%t/%N.cid
Delegate=yes
Type=notify
NotifyAccess=all
SyslogIdentifier=%N
ExecStart=/usr/local/bin/podman run --name=systemd-%N --cidfile=%t/%N.cid --replace --rm --cgroups=split --sdnotify=conmon -d registry.access.redhat.com/ubi9-minimal:latest sleep 1000

---sample_two.service---
[Unit]
Description=The sleep container
After=local-fs.target
SourcePath=/home/debian/.config/containers/systemd/test_subdir/sub_one/sample_two.container
RequiresMountsFor=%t/containers

[X-Container]
Image=registry.access.redhat.com/ubi9-minimal:latest
Exec=sleep 1000

[Install]
# Start by default on boot
WantedBy=multi-user.target default.target

[Service]
Environment=PODMAN_SYSTEMD_UNIT=%n
KillMode=mixed
ExecStop=/usr/local/bin/podman rm -f -i --cidfile=%t/%N.cid
ExecStopPost=-/usr/local/bin/podman rm -f -i --cidfile=%t/%N.cid
Delegate=yes
Type=notify
NotifyAccess=all
SyslogIdentifier=%N
ExecStart=/usr/local/bin/podman run --name=systemd-%N --cidfile=%t/%N.cid --replace --rm --cgroups=split --sdnotify=conmon -d registry.access.redhat.com/ubi9-minimal:latest sleep 1000

---sample_one.service---
[Unit]
Description=The sleep container
After=local-fs.target
SourcePath=/home/debian/.config/containers/systemd/test_subdir/sub_two/sample_one.container
RequiresMountsFor=%t/containers

[X-Container]
Image=registry.access.redhat.com/ubi9-minimal:latest
Exec=sleep 1000

[Install]
# Start by default on boot
WantedBy=multi-user.target default.target

[Service]
Environment=PODMAN_SYSTEMD_UNIT=%n
KillMode=mixed
ExecStop=/usr/local/bin/podman rm -f -i --cidfile=%t/%N.cid
ExecStopPost=-/usr/local/bin/podman rm -f -i --cidfile=%t/%N.cid
Delegate=yes
Type=notify
NotifyAccess=all
SyslogIdentifier=%N
ExecStart=/usr/local/bin/podman run --name=systemd-%N --cidfile=%t/%N.cid --replace --rm --cgroups=split --sdnotify=conmon -d registry.access.redhat.com/ubi9-minimal:latest sleep 1000
```

### CODE REVIEW

**Review by giuseppe — APPROVED** [review]
LGTM

File: cmd/quadlet/main.go
Comment (harikannan512) [reply cmd/quadlet/main.go:112]: I can add a check to see if the path provided by this env var is valid (`filepath.Abs(filePath)`) if that sounds good

File: cmd/quadlet/main.go
Comment (ygalblum) [inline cmd/quadlet/main.go:109]: `unitDirsEnv` is a colon delimited list of directories. So, once the list is longer than 1, `isValidFilePath` will return `false`, no?

File: cmd/quadlet/main.go
Comment (ygalblum) [inline cmd/quadlet/main.go:140]: `Abs` does not validate that the path exists. It guaranties that you get an absolute path by checking if the path is absolute (i.e. starts with `/`) or relative. In the former case, it returns the value as is and in the latter it creates an absolute path by appending the current running directory to the relative path. In both cases, it does not check anything about the validity of the path.
`filepath` also includes an `IsAbs` function that returns `true` if the path is absolute. But, similarly, it does not check that the path exists.

File: cmd/quadlet/main.go
Comment (ygalblum) [reply cmd/quadlet/main.go:116]: @TomSweeneyRedHat we didn't have that before, so I would not require it in this PR. Moreover, what would you check?

File: cmd/quadlet/main.go
Comment (ygalblum) [inline cmd/quadlet/main.go:134]: Why don't you do the same for here?

File: test/e2e/common_test.go
Comment (ygalblum) [inline test/e2e/common_test.go:1421]: This code will return `true` for any error that is not `fs.ErrNotExist`

File: test/e2e/common_test.go
Comment (ygalblum) [inline test/e2e/common_test.go:1433]: According to its [doc](https://pkg.go.dev/os#MkdirAll), `MkdirAll` already noops if the directory exists

**Review by ygalblum — COMMENTED** [review]
This change also requires some documentation. First, about this capability and second about processing order in case of duplications.

File: cmd/quadlet/main.go
Comment (harikannan512) [reply cmd/quadlet/main.go:134]: From what I understood, recursive runs were not required in system quadlets. I was thinking `UnitDirAdmin` would be where admin unit files would go, hence system quadlets be run.
https://github.com/containers/podman/pull/18407#issuecomment-1533507261

File: test/e2e/common_test.go
Comment (harikannan512) [reply test/e2e/common_test.go:1433]: I added the error check to avoid potential golint errors during ci

File: test/e2e/common_test.go
Comment (ygalblum) [inline test/e2e/common_test.go:1342]: Rename `scrDir` -> `srcDir`

File: test/e2e/common_test.go
Comment (ygalblum) [reply test/e2e/common_test.go:1432]: What I meant was that you don't need to call `Exists`

Comment (ygalblum) [reply test/e2e/common_test.go:1420]: Sorry for being a nag. I should have thought about it before. This function now became a wrapper to `os.MkdirAll` that doesn't add anything meaningful to it

Comment (harikannan512) [reply test/e2e/common_test.go:1420]: aah ok, thanks for letting me know. Apologies, I should have noticed it myself

**Comment by ygalblum** [comment]
LGTM
Because there were many changes since @giuseppe's approval, I would like to get another LGTM before merging. @TomSweeneyRedHat @giuseppe can you please take another look?

File: cmd/quadlet/main.go
Comment (giuseppe) [inline cmd/quadlet/main.go:143]: you could just use `%q` instead of `\"%s\"`

File: test/e2e/common_test.go
Comment (giuseppe) [inline test/e2e/common_test.go:1358]: same here

File: test/e2e/common_test.go
Comment (giuseppe) [inline test/e2e/common_test.go:1364]: and here

File: test/e2e/quadlet_test.go
Comment (Luap99) [inline test/e2e/quadlet_test.go:462]:
```suggestion
			GinkgoWriter.Println("QuadletDir: " + quadletDir)
```
never write directly to stdout in the ginkgo tests, this will make the output unreadable.

**Comment by Luap99** [comment]
Also I don't see any changes to the documentation.

**Comment by rhatdan** [comment]
Does this handle rootful containers correctly?

If I put a quadlet in /etc/containers/systemd/users/1000/myquadlet.container, does this get picked up as a system container?  If it does, then that is a bug.  It should only be used for a user with a UID==1000.

**Comment by harikannan512** [comment]
> Does this handle rootful containers correctly?
> 
> If I put a quadlet in /etc/containers/systemd/users/1000/myquadlet.container, does this get picked up as a system container? If it does, then that is a bug. It should only be used for a user with a UID==1000.

**Comment by harikannan512** [comment]
> Also I don't see any changes to the documentation.

@Luap99  I have added a couple of lines to the documentation now. If there is a specific expectation, would be happy to know so that I can add to this

**Comment by ygalblum** [comment]
> Quadlet doesn't run `myquadlet.container` when run as user -> 1000 without `-user`. But when run in sudo or otherwise with `-user`, it does get picked up

Now I'm concerned about the fact that it does not get picked up (not that it should). How do we explain that adding `/etc/containers/systemd` and all its sub-tree does not add `/etc/containers/systemd/users/1000/` to the list of directories?

**Comment by vrothberg** [comment]
> > Quadlet doesn't run `myquadlet.container` when run as user -> 1000 without `-user`. But when run in sudo or otherwise with `-user`, it does get picked up
> 
> Now I'm concerned about the fact that it does not get picked up (not that it should). How do we explain that adding `/etc/containers/systemd` and all its sub-tree does not add `/etc/containers/systemd/users/1000/` to the list of directories?

It shouldn't get picked up.  We probably need to make an exception for the `users` sub-dir?

**Comment by ygalblum** [comment]
> It shouldn't get picked up. We probably need to make an exception for the `users` sub-dir?

I agree that it shouldn't. But, according to @harikannan512 to already doesn't which I don't understand why. This makes me think that unless we can explain it, there's a bug that we're missing somewhere

**Comment by rhatdan** [comment]
If we want to recursive walk directories under /etc/containers/systemd looking for quadlets, then we need to only look for directories under  /etc/containers/systemd/users/$UID where UID==current UID.  IE If the quadlet generator is running as root it should IGNORE quadlets in /etc/containers/systemd/users/ subdirs not starting with UID=0.

File: cmd/quadlet/main.go
Comment (TomSweeneyRedHat) [reply cmd/quadlet/main.go:112]: FWIW, my main concern was that someone might add something like `.../../../../etc` or some such to get into system or other spaces they should not be.  SO a check to make sure something wasn't a relative path would be a starting point.

Comment (ygalblum) [reply cmd/quadlet/main.go:112]: As @harikannan512 wrote, he can add a check that the path is in fact absolute (the method is `IsAbs`). But, I don't understand how would a relative path be different from an absolute path w.r.t. your concern. AFAIK Quadlet is executed with the user's (root or specific user) permissions. So, Quadlet shouldn't be able to access what the user cannot.

Comment (TomSweeneyRedHat) [reply cmd/quadlet/main.go:112]: From past burns, "shouldn't" doesn't always equal "can't".  I'm not an expert on the how-to's in taking advantage of relative references,  I just know that they have often been manipulated badly and I try to avoid them at all costs.

Comment (ygalblum) [reply cmd/quadlet/main.go:112]: NP, so let's at least enforce that is the environment variable is set, it has to be an absolute path

File: cmd/quadlet/main.go
Comment (ygalblum) [reply cmd/quadlet/main.go:140]: I've unresolved this comment since as I wrote here, `Abs` makes an absolute path from a path that is either relative or already absolute.
In fact, [IsAbs](https://pkg.go.dev/path/filepath#IsAbs) is what you're looking for.
Now, please note that if all what this function does is call `IsAbs` you should just call the latter directly without implementing the formar.

**Comment by vrothberg** [comment]
> If we want to recursive walk directories under /etc/containers/systemd looking for quadlets, then we need to only look for directories under /etc/containers/systemd/users/$UID where UID==current UID. IE If the quadlet generator is running as root it should IGNORE quadlets in /etc/containers/systemd/users/ subdirs not starting with UID=0.

Yes, that sounds reasonable.

**Comment by ygalblum** [comment]
@harikannan512 I see that you make sure the PR is in tact by rebasing. But, we in order to merge this change, we need to resolve the discussion about the users directories under `/etc/containers/systemd`

**Comment by harikannan512** [comment]
> @harikannan512 I see that you make sure the PR is in tact by rebasing. But, we in order to merge this change, we need to resolve the discussion about the users directories under `/etc/containers/systemd`

Sure thing. Just to reiterate my understanding - when quadlet is run as sudo, it should scan `/etc/containers/systemd` and `/usr/share/containers/systemd` directories and when quadlet runs as sudo with `-user` option, it should be able to scan sub dirs within directories defined by root UIDs (i.e. 0*), but not the other user directories.
image
image

**Comment by ygalblum** [comment]
@harikannan512 I think that first, we need to keep in mind that Quadlet is a systemd generator.

If we set aside `dry-run` Quadlet is executed either via `/usr/lib/systemd/system-generators/podman-system-generator` or `/usr/lib/systemd/user-generators/podman-user-generator` without the `-user` parameter in either case. Quadlet figures out the context using the program name [here](https://github.com/containers/podman/blob/main/cmd/quadlet/main.go#L321), though it might have been preferable to use the [SYSTEMD_SCOPE](https://www.freedesktop.org/software/systemd/man/systemd.generator.html#%24SYSTEMD_SCOPE) environment variable. In any case, the option of running as root with `-user` set is not real life use case.

Now, let's look at the current implementation of [getUnitDirs](https://github.com/containers/podman/blob/main/cmd/quadlet/main.go#L102). We can see that for `root` it adds two directories: `/etc/containers/systemd` and `/usr/share/containers/systemd` while for rootless, it adds `~/.config/containers/systemd`, `/etc/containers/systemd/users` and `/etc/containers/systemd/users/$UID`.

So, as I see it that are two possible issues:
1. When running as `root` we want to traverse the sub-dirs of `/etc/containers/systemd` but we need to make sure that we don't go into the `users` directory as this sub-tree includes units for user level services that should not be executed for `root`.
2. When running as non-root we need to make sure that when traversing `/etc/containers/systemd/users` we don't go into other users' directory. This one, I guess is a bit more tricky because there is no one directory not to enter to. Maybe enter all directories that are not a number unless the number is the user's UID?

@rhatdan please confirm my analysis to make sure we're not redoing this section over and over again.

**Comment by rhatdan** [comment]
Yes I agree with your analysys.

**Comment by harikannan512** [comment]
I believe I confused the system superuser with Podman root/admin user. These new changes ensure 

a) If root, quadlet will only run user files outside of `users` folder, and will NOT enter `users` dir, even for user id 0.
image

b) When rootless, the users folder will be traverses recursively, but only for 
- the current user id
- non numeric sub dirs
User ID: 1000
image

User ID: 1001
image

File: cmd/quadlet/main.go
Comment (ygalblum) [inline cmd/quadlet/main.go:145]: `appendSubPaths` and `appendNonNumericSubPaths` are practically the same with the sole difference of a filter method.
Consider combining them into one function that get a `filter` function pointer (that can be `nil`)

File: cmd/quadlet/main.go
Comment (ygalblum) [inline cmd/quadlet/main.go:146]: Consider using [WalkDir](https://pkg.go.dev/path/filepath#WalkDir) as it is more efficient

File: cmd/quadlet/main.go
Comment (ygalblum) [inline cmd/quadlet/main.go:194]: I know your code first checks that it's under `users` and calls `SystemUserDirsExist` which guaranties that `listDirUserPaths` is long enough. But, functions (and their usage) tend to take a life of their own.
So, I think you should either not have them as separate methods or make them more tight in terms of what they check

File: cmd/quadlet/main.go
Comment (ygalblum) [inline cmd/quadlet/main.go:148]: `filterPtr` should just return a `boolean` if the directory should be appended. This way the `if` here can be:
```
if filterPtr == nil || filterPtr(_path, isUserFlag) {
  dirs = append(dirs, _path)
}
```

**Review by ygalblum — COMMENTED** [review]
LGTM

**Comment by ygalblum** [comment]
@harikannan512 you'll need to rebase on top of main in order to take in the fix for the Windows installer

**Comment by Luap99** [comment]
just a drive by comment but please squash the commits into one before merging

**Comment by ygalblum** [comment]
@rhatdan @vrothberg PTAL

Also, was the Rawhide issue resolved? I see many PRs failing on it

**Comment by Luap99** [comment]
> Also, was the Rawhide issue resolved? I see many PRs failing on it

https://github.com/containers/gvisor-tap-vsock/issues/232
Ignore it for now it is not blocking CI.

**Comment by rhatdan** [comment]
/lgtm
@harikannan512 Thanks for sticking with this.

**Pull request merged**
